### PR TITLE
Make Globalize thread safe by default again.

### PR DESCRIPTION
**Pull request by simi**
I think it is easy enough to use custom storage just by overriding storage method on Globalize. I was thinking about getter and setter for that, but proc will be needed there probably and it looks ugly in config.

https://github.com/globalize/globalize/commit/37c19b5f8b19ba14bc89f7237064857d46ad815d :red_circle: fails without https://github.com/globalize/globalize/commit/b1857194a147ef2a0d3bd99d3c7cab17d0683edc :green_heart: 

related to #390 #374 

ping @mtparet @hube @parndt @Dschee

### CODE REVIEW

File: lib/globalize.rb [hunk 1]
```diff
@@ -73,7 +76,7 @@ def set_fallbacks(locales)
         fallback_hash[key] = value.presence || [key]
       end if locales.present?
 
-      @fallbacks = fallback_hash
+      Thread.current[:globalize_fallbacks] = fallback_hash
```
Comment (parndt) [inline lib/globalize.rb]: This can't use storage?

Comment (simi) [inline lib/globalize.rb]: sec

Comment (simi) [inline lib/globalize.rb]: good catch @parndt, fixed

**Comment by mtparet** [comment]
I didn't take the time for this, thank you @simi ! :heart:

File: test/globalize_test.rb [hunk 2]
```diff
@@ -217,4 +217,28 @@ class GlobalizeTest < MiniTest::Spec
       end
     end
   end
+
+  describe 'with locale' do
+    def perform_with_locale(locale)
+      Thread.new do
+        Globalize.with_locale(locale) do
+          sleep 1
```
Comment (parndt) [inline test/globalize_test.rb]: Oh no, this will make our tests ~75% slower! :wink:

Comment (mtparet) [inline test/globalize_test.rb]: 0.1 should be enough :)

Comment (simi) [inline test/globalize_test.rb]: Actually not. I tested it.
Dne 16.2.2015 1:22 "Matthieu Paret" <notifications@github.com> napsal(a):

Comment (mtparet) [inline test/globalize_test.rb]: :-( it depends of the system.

Comment (simi) [inline test/globalize_test.rb]: I have really poor one.
Dne 16.2.2015 1:41 "Matthieu Paret" <notifications@github.com> napsal(a):

**Comment by scarfacedeb** [comment]
Maybe, we should use [request_store](https://github.com/steveklabnik/request_store), instead of Thread.current to prevent an unwanted sharing of values in multiple requests?
I'm not a thread expert, but it seems sane.

**Comment by simi** [comment]
@scarfacedeb RequestStore is meant to be used in Rack application as Rack middleware. So it is not usable in Sidekiq for example.

**Comment by scarfacedeb** [comment]
@simi OK. 
I've found an old issue (https://github.com/globalize/globalize/issues/224) about that kind of unwanted behaviour and @parndt also suggested to use request_store.

**Comment by parndt** [comment]
@steveklabnik hey, would you recommend using request_store for this?

**Comment by steveklabnik** [comment]
Yes, this code is effectively RequestStore.

@simi while it's true that I concieved it as being out of Rack, there's no reason why you couldn't use it in Sidekiq. You'd just need to make sure to call the clear function at the right time. I don't use Sidekiq, so I don't know what kind of hooks it has, off the top of my head.

**Comment by parndt** [comment]
Thanks, @steveklabnik !

@mperham do you have any wisdom here? Would it be conceiveable that this would work with Sidekiq?

**Comment by steveklabnik** [comment]
https://github.com/steveklabnik/request_store/blob/master/lib/request_store/middleware.rb#L7-L11 <- you'd need to call that method (`clear!`) on startup. given https://github.com/mperham/sidekiq/wiki/Middleware seems fine.

I would not mind having said middleware inside of the gem itself, if that's helpful.

**Comment by mperham** [comment]
Here's Sidekiq's i18n middleware: https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/middleware/i18n.rb  Yeah, `Thread.current` is totally cool with Sidekiq; I use it everywhere.

**Comment by simi** [comment]
This is not just about Sidekiq, but about non-rack env globally.

What about to stick with `Thread.current` and add `Globalize.clear!` method to allow manual clear when you'll need it. For example in Rails controller, you can use it in `after_filter`.

**Comment by parndt** [comment]
Thanks, @mperham !

**Comment by scarfacedeb** [comment]
@simi Wouldn't it basically duplicate RequestStore?
What if we use RequestStore which would provide automatic clearing on every request.
And non-Rack apps could call RequestStore.clear!

That way the most popular case (in a rails app) is getting better thread safety for free.

**Comment by simi** [comment]
That makes sense @scarfacedeb. Would you mind to send PR for that?

**Comment by scarfacedeb** [comment]
@simi Sure! I'll look into it.

File: test/globalize_test.rb (hunk 2 above)
Comment (scarfacedeb) [inline test/globalize_test.rb]: @simi Could you clarify why is `sleep` needed here?

Is it to give a scheduler time to switch into another thread?
Maybe we could use `Thread.pass` instead

Comment (simi) [inline test/globalize_test.rb]: Just copy paste from https://github.com/globalize/globalize/issues/390#issue-52729253. Feel free to make red test (revert my patch - use ivars as storage) and then make it green with `RequestStore`.

Comment (scarfacedeb) [inline test/globalize_test.rb]: This test is green with or without `sleep` on my machine. That's why I thought that maybe I'm missing something.
I'll use it without `sleep` then, until there's a reason to have it there.

Comment (simi) [inline test/globalize_test.rb]: Try to run it with sleep and ivars storage.

Comment (scarfacedeb) [inline test/globalize_test.rb]: I see what you mean.
With ivar and sleep: it fails, but when I remove the sleep - it suddenly passes.
The best news is that it also fails with `Thread.pass` and ivars, but passes with Thread.current.
I suppose it means that we can replace sleep with Thread.pass :)

Comment (simi) [inline test/globalize_test.rb]: Good catch! Let's use `Thread.pass`.

**Comment by simi** [comment]
#420 done it.

**Pull request closed without merging**